# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1709828394 -3600
# Parent  1f09aae39d34bca8521411938ecd781a2ac11b65
Feature: Account colours, part 2: Message list.
* * *
Introduce pref mail.threadpane.full_row_color to allow full row color.

diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -367,8 +367,10 @@ pref("mail.threadpane.padding.bottom_per
 
 // Use correspondents column instead of from/recipient columns.
 pref("mail.threadpane.use_correspondents", true);
 
+pref("mail.threadpane.full_row_color", false);
+
 // To allow images to be inserted into a composition with an auth prompt, we
 // need the following two.
 pref("network.auth.subresource-img-cross-origin-http-auth-allow", true);
 pref("network.auth.non-web-content-triggered-resources-http-auth-allow", true);
diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -790,8 +790,10 @@ var folderPane = {
    * @type {boolean}
    */
   _hideLocalFolders: false,
 
+  _serverColorMap: null,
+
   _modes: {
     all: {
       name: "all",
       active: false,
@@ -2636,8 +2638,14 @@ var folderPane = {
       } else if (f.isServer) {
         row.style.backgroundColor = color;
       }
     }
+
+    // Take care of cached values.
+    if (this._serverColorMap) {
+      this._serverColorMap.delete(folder.server.key);
+    }
+    threadTree.invalidate();
   },
 
   /**
    * Update the UI widget to reflect the real folder size when the "FolderSize"
@@ -4664,14 +4672,20 @@ var threadPane = {
   columns: ThreadPaneColumns.getDefaultColumns(gFolder),
 
   cardColumns: ThreadPaneColumns.getDefaultColumnsForCardsView(gFolder),
 
+  fullRowColor: Services.prefs.getBoolPref(
+    "mail.threadpane.full_row_color",
+    false
+  ),
+
   async init() {
     quickFilterBar.init();
 
     this.setUpTagStyles();
     Services.prefs.addObserver("mailnews.tags.", this);
     Services.prefs.addObserver("mail.pane_config.multiline_all", this);
+    Services.prefs.addObserver("mail.threadpane.full_row_color", this);
 
     Services.obs.addObserver(this, "addrbook-displayname-changed");
     Services.obs.addObserver(this, "custom-column-added");
     Services.obs.addObserver(this, "custom-column-removed");
@@ -4858,8 +4872,14 @@ var threadPane = {
           );
           customElements.get("thread-row").MULTILINE = multiline;
           this.densityChange();
           threadTree.reset();
+        } else if (data == "mail.threadpane.full_row_color") {
+          this.fullRowColor = Services.prefs.getBoolPref(
+            "mail.threadpane.full_row_color",
+            false
+          );
+          threadTree.reset();
         } else {
           this.setUpTagStyles();
         }
         break;
@@ -6666,8 +6686,40 @@ var folderListener = {
     }
   },
 };
 
+function getRGBAcolorForServerKey(key) {
+  // Create the map if we don't have one.
+  if (!folderPane._serverColorMap) {
+    folderPane._serverColorMap = new Map();
+  }
+
+  // Return the rgba if we have it in the map.
+  let rgba = folderPane._serverColorMap.get(key);
+  if (rgba !== undefined) {
+    return rgba;
+  }
+
+  // Determine the rgba and put it into the map.
+  let color = Services.prefs.getCharPref(`mail.server.${key}.color`, "");
+  if (
+    !color ||
+    color == "#000000" ||
+    !color.startsWith("#") ||
+    color.length != 7
+  ) {
+    rgba = "";
+  } else {
+    // Turn #RRGGBB into rgba(rr, gg, bb, tt)
+    let r = parseInt(color.substring(1, 3), 16);
+    let g = parseInt(color.substring(3, 5), 16);
+    let b = parseInt(color.substring(5, 7), 16);
+    rgba = `rgba(${r}, ${g}, ${b}, 0.2)`;
+  }
+  folderPane._serverColorMap.set(key, rgba);
+  return rgba;
+}
+
 /**
  * Custom element for rows in the thread tree.
  */
 customElements.whenDefined("tree-view-table-row").then(() => {
@@ -6722,31 +6774,57 @@ customElements.whenDefined("tree-view-ta
       // XPCOM calls here must be keep to a minimum. Collect all of the
       // required data in one go.
       let threadedDisplay =
         gViewWrapper._viewFlags & Ci.nsMsgViewFlagsType.kThreadedDisplay;
+      let getServerInfo = !gViewWrapper.isSingleFolder;
       let properties = {};
       let threadLevel = {};
       let numChildren = {};
       let hasNextSibling = {};
+      let serverKey = {};
       let cellTexts = this.view.cellDataForColumns(
         index,
         textColumns,
         properties,
         threadLevel,
         threadedDisplay, // no need to get the following two for a non-threaded display
         numChildren,
-        hasNextSibling
+        hasNextSibling,
+        getServerInfo,
+        serverKey
       );
+      const propertiesSet = new Set(properties.value.split(" "));
+      const isDummyRow = propertiesSet.has("dummy");
+
+      let elementToColor;
+      if (threadPane.fullRowColor) {
+        elementToColor = this;
+      } else {
+        // In multi-line mode, don't allow to color the displaced subject column.
+        let selector = ThreadRow.MULTILINE
+          ? "td:not([hidden]):not(.subjectcol-column)"
+          : "td:not([hidden])";
+        elementToColor = this.querySelector(selector);
+      }
+      if (elementToColor) {
+        if (getServerInfo && !isDummyRow) {
+          let color = getRGBAcolorForServerKey(serverKey.value);
+          if (color) {
+            elementToColor.style.backgroundColor = color;
+          } else {
+            elementToColor.style.backgroundColor = "";
+          }
+        } else {
+          elementToColor.style.backgroundColor = "";
+        }
+      }
       this.classList.toggle("singleton", numChildren.value <= 1);
       this.classList.toggle("lastchild", !hasNextSibling.value);
 
       // Collect the various strings and fluent IDs to build the full string for
       // the message row aria-label.
       let ariaLabelPromises = [];
 
-      const propertiesSet = new Set(properties.value.split(" "));
-      const isDummyRow = propertiesSet.has("dummy");
-
       this.dataset.properties = properties.value.trim();
 
       let seenSubjectColumn = false;
       for (const column of columns) {
@@ -6996,31 +7074,61 @@ customElements.whenDefined("tree-view-ta
       super.index = index;
 
       // XPCOM calls here must be keep to a minimum. Collect all of the
       // required data in one go.
+      let getServerInfo = !gViewWrapper.isSingleFolder;
       let properties = {};
       let threadLevel = {};
       let numChildren = {};
       let hasNextSibling = {};
+      let serverKey = {};
 
       let cellTexts = this.view.cellDataForColumns(
         index,
         threadPane.cardColumns,
         properties,
         threadLevel,
         false,
         numChildren,
-        hasNextSibling
+        hasNextSibling,
+        getServerInfo,
+        serverKey
       );
+      if (threadLevel.value) {
+        properties.value += " thread-children";
+      }
+      const propertiesSet = new Set(properties.value.split(" "));
+      const isDummyRow = propertiesSet.has("dummy");
+
+      if (threadPane.fullRowColor) {
+        if (getServerInfo && !isDummyRow) {
+          let color = getRGBAcolorForServerKey(serverKey.value);
+          if (color) {
+            this.style.backgroundColor = color;
+          } else {
+            this.style.backgroundColor = "";
+          }
+        } else {
+          this.style.backgroundColor = "";
+        }
+      } else {
+        // eslint-disable-next-line no-lonely-if
+        if (getServerInfo && !isDummyRow) {
+          let color = getRGBAcolorForServerKey(serverKey.value);
+          if (color) {
+            this.firstElementChild.style.borderLeft = `5px solid ${color}`;
+          } else {
+            this.firstElementChild.style.borderLeft = "";
+          }
+        } else {
+          this.firstElementChild.style.borderLeft = "";
+        }
+      }
 
       // Collect the various strings and fluent IDs to build the full string for
       // the message row aria-label.
       let ariaLabelPromises = [];
 
-      if (threadLevel.value) {
-        properties.value += " thread-children";
-      }
-      const propertiesSet = new Set(properties.value.split(" "));
       this.dataset.properties = properties.value.trim();
 
       this.subjectLine.textContent = cellTexts[0];
       this.subjectLine.title = cellTexts[0];
diff --git a/mailnews/base/public/nsIMsgDBView.idl b/mailnews/base/public/nsIMsgDBView.idl
--- a/mailnews/base/public/nsIMsgDBView.idl
+++ b/mailnews/base/public/nsIMsgDBView.idl
@@ -496,8 +496,10 @@ interface nsIMsgDBView : nsISupports
    * @param aThreadLevel - The thread level of the row.
    * @param aGetChildrenInfo - if false, return dummy values for aNumChildren and aHasNextSibling.
    * @param aNumChildren - The number of children of the row.
    * @param aHasNextSibling - Whether the row has a next sibling.
+   * @param aGetServerInfo - if false, return dummy value for aServerKey.
+   * @param aServerKey - Key of the server belonging to the folder.
    *
    * @returns The cell text for the columns in `aColumnNames`.
    */
   Array<AString> cellDataForColumns(in long aRow,
@@ -505,9 +507,11 @@ interface nsIMsgDBView : nsISupports
                                     out AString aProperties,
                                     out long aThreadLevel,
                                     in boolean getChildrenInfo,
                                     out unsigned long aNumChildren,
-                                    out boolean aHasNextSibling);
+                                    out boolean aHasNextSibling,
+                                    in boolean aGetServerInfo,
+                                    out ACString aServerKey);
 };
 
 /* this interface is rapidly morphing from a command updater interface into a more generic
    FE updater interface to handle changes in the view
diff --git a/mailnews/base/src/nsMsgDBView.cpp b/mailnews/base/src/nsMsgDBView.cpp
--- a/mailnews/base/src/nsMsgDBView.cpp
+++ b/mailnews/base/src/nsMsgDBView.cpp
@@ -1983,9 +1983,10 @@ NS_IMETHODIMP
 nsMsgDBView::CellDataForColumns(int32_t aRow,
                                 const nsTArray<nsString>& aColumnNames,
                                 nsAString& aProperties, int32_t* aThreadLevel,
                                 bool aGetChildrenInfo, uint32_t* aNumChildren,
-                                bool* aHasNextSibling,
+                                bool* aHasNextSibling, bool aGetServerInfo,
+                                nsACString& aServerKey,
                                 nsTArray<nsString>& _retval) {
   nsresult rv;
   _retval.Clear();
 
@@ -2023,11 +2024,25 @@ nsMsgDBView::CellDataForColumns(int32_t 
   rv = GetLevel(aRow, aThreadLevel);
   if (NS_FAILED(rv)) {
     _retval.Clear();
     aProperties.Truncate();
-  }
-
-  return rv;
+    return rv;
+  }
+
+  aServerKey.Truncate();
+  if (!aGetServerInfo) return NS_OK;
+
+  nsCOMPtr<nsIMsgDBHdr> msgHdr;
+  rv = GetMsgHdrForViewIndex(aRow, getter_AddRefs(msgHdr));
+  NS_ENSURE_SUCCESS(rv, NS_OK);
+  nsCOMPtr<nsIMsgFolder> folder;
+  msgHdr->GetFolder(getter_AddRefs(folder));
+  if (folder) {
+    nsCOMPtr<nsIMsgIncomingServer> server;
+    folder->GetServer(getter_AddRefs(server));
+    if (server) server->GetKey(aServerKey);
+  }
+  return NS_OK;
 }
 
 NS_IMETHODIMP
 nsMsgDBView::SetTree(mozilla::dom::XULTreeElement* tree) {
